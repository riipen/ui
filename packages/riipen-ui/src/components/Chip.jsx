import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class Chip extends React.Component {
  static propTypes = {
    /**
     * The elements to render inside the pill.
     */
    children: PropTypes.node,

    /**
     * List of additional classes to apply to this component.
     */
    classes: PropTypes.array,

    /**
     * The color to use.
     */
    color: PropTypes.oneOf([
      "default",
      "dark",
      "primary",
      "secondary",
      "tertiary",
      "positive",
      "warning",
      "negative"
    ]),

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    component: PropTypes.elementType,

    /**
     * If `true`, the chip should be displayed in a disabled state.
     */
    disabled: PropTypes.bool,

    /**
     * Icon to display at start of chip.
     */
    icon: PropTypes.elementType,

    /**
     * The content of the label.
     */
    label: PropTypes.node,

    /**
     * Action to perform when clicked.
     */
    onClick: PropTypes.func,

    /**
     * Callback function fired when the icon is clicked.
     */
    onIconClick: PropTypes.func,

    /**
     * The size of the chip.
     */
    size: PropTypes.oneOf(["small", "medium"]),

    /**
     * The variant to use.
     */
    variant: PropTypes.oneOf(["default", "outlined"])
  };

  static defaultProps = {
    classes: [],
    color: "default",
    component: "div",
    disabled: false,
    size: "medium",
    variant: "default"
  };

  static contextType = ThemeContext;

  render() {
    const {
      classes,
      color,
      component: Component,
      disabled,
      icon: Icon,
      onIconClick,
      onClick,
      size,
      variant
    } = this.props;

    const label = this.props.label || this.props.children;

    const theme = this.context;

    const clickable = !!onClick;
    const iconClickable = !!onIconClick;

    const className = clsx(
      "root",
      color,
      disabled ? "disabled" : null,
      size,
      variant,
      classes,
      { clickable }
    );

    const handleClick = () => {
      if (onClick) onClick();
    };

    return (
      <React.Fragment>
        <Component onClick={handleClick} className={className}>
          {Icon && (
            <span
              onClick={onIconClick}
              className={clsx("icon", "icon-start", {
                iconClickable
              })}
            >
              <Icon />
            </span>
          )}
          <span className={clsx("label")}>{label}</span>
        </Component>
        <style jsx>{`
          .root {
            background-color: ${theme.palette.grey[300]};
            border-radius: 50px;
            border: 1px solid;
            border-color: transparent;
            box-sizing: border-box;
            color: ${theme.palette.text.secondary};
            cursor: default;
            display: inline-flex;
            font-family: ${theme.typography.fontFamily};
            font-size: 13px;
            justify-content: center;
            outline: 0;
            vertical-align: middle;
            white-space: nowrap;
          }
          .root.small {
            padding: 5px 10px;
          }
          .root.medium {
            padding: 10px 15px;
          }
          .root.disabled {
            background-color: ${theme.palette.disabled};
            border: 1px solid;
            border-color: ${theme.palette.disabled};
            color: ${theme.palette.common.white};
            pointer-events: none;
          }
          .root.clickable {
            cursor: pointer;
          }

          .icon-start {
            align-items: center;
            display: flex;
            padding-right: 10px;
            white-space: nowrap;
          }

          .icon-start.disabled {
            cursor: initial;
          }

          .icon-start.iconClickable {
            cursor: pointer;
          }

          .label {
            align-items: center;
            display: flex;
            white-space: nowrap;
          }

          .primary {
            background-color: ${theme.palette.primary.main};
            border-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.contrast};
          }
          .primary.outlined {
            border-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.main};
          }
          .primary.outlined.hover:hover {
            background-color: ${theme.palette.primary.main};
            border-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.contrast};
          }

          .secondary {
            background-color: ${theme.palette.secondary.main};
            border-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.contrast};
          }
          .secondary.outlined {
            border-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.main};
          }
          .secondary.outlined.hover:hover {
            background-color: ${theme.palette.secondary.main};
            border-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.contrast};
          }

          .tertiary {
            background-color: ${theme.palette.tertiary.main};
            border-color: ${theme.palette.tertiary.main};
            color: ${theme.palette.tertiary.contrast};
          }
          .tertiary.outlined {
            border-color: ${theme.palette.tertiary.main};
            color: ${theme.palette.tertiary.main};
          }

          .negative {
            background-color: ${theme.palette.negative.main};
            border-color: ${theme.palette.negative.main};
            color: ${theme.palette.negative.contrast};
          }
          .negative.outlined {
            border-color: ${theme.palette.negative.main};
            color: ${theme.palette.negative.main};
          }

          .warning {
            background-color: ${theme.palette.warning.main};
            border-color: ${theme.palette.warning.main};
            color: ${theme.palette.warning.contrast};
          }
          .warning.outlined {
            border-color: ${theme.palette.warning.main};
            color: ${theme.palette.warning.main};
          }

          .positive {
            background-color: ${theme.palette.positive.main};
            border-color: ${theme.palette.positive.main};
            color: ${theme.palette.positive.contrast};
          }
          .positive.outlined {
            border-color: ${theme.palette.positive.main};
            color: ${theme.palette.positive.main};
          }

          .dark {
            background-color: ${theme.palette.grey.A400};
            border-color: ${theme.palette.grey.A400};
            color: ${theme.palette.common.white};
          }
          .dark.outlined {
            border-color: ${theme.palette.grey.A400};
            color: ${theme.palette.grey.A400};
          }

          .outlined {
            background-color: transparent;
            border: 1px solid ${theme.palette.text.primary};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(Chip);
