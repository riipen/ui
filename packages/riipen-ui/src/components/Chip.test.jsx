import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Chip from "./Chip";

describe("<Chip>", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders without errors", () => {
    let error;
    try {
      mount(<Chip />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Chip().type.defaultProps;

      const wrapper = mount(<Chip />);

      const component = wrapper.find("Chip").childAt(0);
      expect(component.name()).toEqual("div");
      expect(component.hasClass(defaultProps.color)).toEqual(true);
      expect(component.hasClass("disabled")).toEqual(false);
      expect(component.hasClass("hover")).toEqual(false);
      expect(component.hasClass(defaultProps.size)).toEqual(true);
      expect(component.hasClass(defaultProps.variant)).toEqual(true);
      expect(component.childAt(0).name()).toEqual("span");

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe("children prop", () => {
    it("displays label from child", () => {
      const label = "A label given as a child";
      const child = <div>{label}</div>;

      const wrapper = mount(<Chip>{child}</Chip>);

      expect(
        wrapper
          .find(".label")
          .children()
          .contains(child)
      ).toEqual(true);
      expect(
        wrapper
          .find(".label")
          .childAt(0)
          .text()
      ).toEqual(label);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the root node of Chip", () => {
      const classes = ["classOne", "classTwo"];

      const wrapper = mount(<Chip classes={classes} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(classes[1])
      ).toEqual(true);
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Chip classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("sets color class name with a valid color", () => {
      const color = "dark";

      const wrapper = mount(<Chip color={color} />);

      expect(wrapper.find("Chip").props().color).toEqual(color);
      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(color)
      ).toEqual(true);
      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(classes[1])
      ).toEqual(true);
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Chip classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("sets color class name with a valid color", () => {
      const color = "dark";

      const wrapper = mount(<Chip color={color} />);

      expect(wrapper.find("Chip").props().color).toEqual(color);
      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(color)
      ).toEqual(true);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Chip color="red" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Chip color="red" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("component prop", () => {
    it("sets component with valid custom component", () => {
      const component = "a";

      const wrapper = mount(<Chip component={component} />);

      expect(wrapper.find(component)).toHaveLength(1);
      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .name()
      ).toEqual(component);
    });

    it("gives an error when given an invalid component", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const component = "abcdefg";

      mount(<Chip component={component} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("disabled prop", () => {
    it("applies disabled class when disabled prop is true", () => {
      const disabled = true;

      const wrapper = mount(<Chip disabled={disabled} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass("disabled")
      ).toEqual(true);
    });

    it("does not apply disabled class when disabled prop is false", () => {
      const disabled = false;

      const wrapper = mount(<Chip disabled={disabled} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass("disabled")
      ).toEqual(false);
    });
  });

  describe("hover prop", () => {
    it("applies hover class when hover prop is true", () => {
      const hover = true;

      const wrapper = mount(<Chip hover={hover} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass("hover")
      ).toEqual(true);
    });

    it("does not apply hover class when hover prop is false", () => {
      const hover = false;

      const wrapper = mount(<Chip hover={hover} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass("hover")
      ).toEqual(false);
    });
  });

  describe("icon prop", () => {
    it("displays icon with valid custom elementType", () => {
      const icon = "a";

      const wrapper = mount(<Chip icon={icon} />);

      expect(wrapper.find(icon)).toHaveLength(1);
      expect(
        wrapper
          .find(".icon")
          .childAt(0)
          .name()
      ).toEqual(icon);
    });

    it("gives an error given an invalid icon type", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const icon = "aaaaaaaaa";

      mount(<Chip icon={icon} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("label prop", () => {
    it("displays label from label prop", () => {
      const label = "A test label";

      const wrapper = mount(<Chip label={label} />);

      expect(
        wrapper
          .find("Chip")
          .find("span")
          .text()
      ).toEqual(label);
    });

    it("prioritizes displaying label over children", () => {
      const label = "label";
      const child = <div>"child"</div>;

      const wrapper = mount(<Chip label={label}>{child}</Chip>);

      expect(
        wrapper
          .find(".label")
          .children()
          .contains(child)
      ).toEqual(false);
      expect(
        wrapper
          .find("Chip")
          .find("span")
          .text()
      ).toEqual(label);
    });
  });

  describe("onClick prop", () => {
    it("invokes onClick handler when clicked", () => {
      const handler = jest.fn();

      const wrapper = mount(<Chip onClick={handler} />);

      wrapper.find("Chip").simulate("click");

      expect(handler).toHaveBeenCalledTimes(1);
    });
  });

  describe("size prop", () => {
    it("sets size class name with a valid size", () => {
      const size = "small";

      const wrapper = mount(<Chip size={size} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(size)
      ).toEqual(true);
    });

    it("gives an error when given an invalid size", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const size = "large";

      mount(<Chip size={size} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("variant prop", () => {
    it("sets variant class name with a valid variant", () => {
      const variant = "outlined";

      const wrapper = mount(<Chip variant={variant} />);

      expect(
        wrapper
          .find("Chip")
          .childAt(0)
          .hasClass(variant)
      ).toEqual(true);
    });

    it("gives an error when given an invalid variant", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const variant = "somethingInvalid";

      mount(<Chip variant={variant} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
