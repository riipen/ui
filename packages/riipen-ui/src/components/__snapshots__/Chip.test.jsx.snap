// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chip> renders correct snapshot 1`] = `
<withClasses(Chip)>
  <Chip
    classes={
      Array [
        "riipen",
        "riipen-chip",
      ]
    }
    color="default"
    component="div"
    disabled={false}
    hover={false}
    size="medium"
    variant="default"
  >
    <div
      className="jsx-3261788765 root default medium default riipen riipen-chip"
      onClick={[Function]}
    >
      <span
        className="jsx-3261788765 label"
      />
    </div>
    <JSXStyle
      dynamic={
        Array [
          "#dddddd",
          "#747474",
          "Roboto, Helvetica, Arial, sans-serif",
          "#cccbcb",
          "#cccbcb",
          "#fff",
          "#373737",
          "#1f71ae",
          "#1f71ae",
          "#fff",
          "#1f71ae",
          "#1f71ae",
          "#1f71ae",
          "#1f71ae",
          "#1f71ae",
          "#1f71ae",
          "#fff",
          "#f58632",
          "#f58632",
          "#fff",
          "#f58632",
          "#f58632",
          "#f58632",
          "#f58632",
          "#f58632",
          "#f58632",
          "#fff",
          "#3caabb",
          "#3caabb",
          "#fff",
          "#3caabb",
          "#3caabb",
          "#3caabb",
          "#3caabb",
          "#3caabb",
          "#3caabb",
          "#fff",
          "#e95353",
          "#e95353",
          "#fff",
          "#e95353",
          "#e95353",
          "#e95353",
          "#e95353",
          "#e95353",
          "#e95353",
          "#fff",
          "#febb46",
          "#febb46",
          "#fff",
          "#febb46",
          "#febb46",
          "#febb46",
          "#febb46",
          "#febb46",
          "#febb46",
          "#fff",
          "#4cc88d",
          "#4cc88d",
          "#fff",
          "#4cc88d",
          "#4cc88d",
          "#4cc88d",
          "#4cc88d",
          "#4cc88d",
          "#4cc88d",
          "#fff",
          "#373737",
          "#373737",
          "#fff",
          "#373737",
          "#373737",
          "#373737",
          "#373737",
          "#373737",
          "#373737",
          "#fff",
          "#373737",
          "#dddddd",
          "#dddddd",
          "#747474",
        ]
      }
      id="3718248004"
    />
  </Chip>
</withClasses(Chip)>
`;
